File: packages/sillytavern-compat/src/support.ts
import type { JsonObject, JsonValue } from "@zhuangsheng/api-client";
import { cloneJson, object, string } from "./json";
import type {
  SillyTavernImportWarningView,
  SillyTavernPresetKind,
  SillyTavernTextTransformView,
  TextTransformRuleView,
} from "./types";

export interface ImportParts {
  contextSpec: JsonObject | null;
  generation: JsonObject | null;
  providerExtensions: JsonObject | null;
  views: SillyTavernTextTransformView[];
  inactiveFields: string[];
  warnings: SillyTavernImportWarningView[];
}

export const PRIORITY = { global: 0, character: 1_000, preset: 2_000 } as const;
const SECTION_KINDS: Record<string, SillyTavernPresetKind> = {
  context: "context",
  instruct: "instruct",
  sysprompt: "system_prompt",
  preset: "text_completion",
  reasoning: "reasoning",
};

export function parts(baseSpec?: JsonObject | null): ImportParts {
  const contextSpec = baseSpec ? cloneJson(baseSpec) : null;
  const rules = Array.isArray(contextSpec?.textTransforms)
    ? contextSpec.textTransforms.flatMap((value) => runtimeRule(value) ? [viewFromRuntime(runtimeRule(value)!)] : [])
    : [];
  return { contextSpec, generation: null, providerExtensions: null, views: rules, inactiveFields: [], warnings: [] };
}

export function emptySpec(name: string): JsonObject {
  return {
    id: null, name, mode: "chat", items: [], budget: null, postProcess: [],
    textTransforms: [], textTransformMacros: {}, preview: null,
  };
}

export function importName(document: JsonValue, sourceName?: string | null) {
  return (string(object(document)?.name) ?? sourceName ?? "SillyTavern import").trim().slice(0, 200);
}

export function warn(parts: ImportParts, code: string, message: string, field: string | null = null) {
  parts.warnings.push({ code, message, field });
}

export function inactive(parts: ImportParts, field: string, message: string) {
  parts.inactiveFields.push(field);
  warn(parts, "sillytavern_field_inactive", message, field);
}

export function installRules(parts: ImportParts, imported: SillyTavernTextTransformView[]) {
  for (const rule of imported) {
    const index = parts.views.findIndex((value) => value.id === rule.id);
    if (index >= 0) parts.views[index] = rule; else parts.views.push(rule);
  }
  parts.views.sort((left, right) => left.priority - right.priority || left.order - right.order);
  const spec = parts.contextSpec ??= emptySpec("SillyTavern regex");
  spec.textTransforms = parts.views.map(stripView) as unknown as JsonValue;
}

export function populateRoleMacros(spec: JsonObject) {
  const macros = object(spec.textTransformMacros) ?? {};
  const items = Array.isArray(spec.items) ? spec.items : [];
  for (const raw of items) {
    const item = object(raw); const id = string(item?.id); const name = string(item?.name);
    if (!id || !name?.trim()) continue;
    const profile = id.split(/[:/]/, 1)[0];
    if (profile === "character") { macros.char ??= name; macros.name2 ??= name; }
    if (profile === "persona") { macros.user ??= name; macros.name1 ??= name; }
  }
  spec.textTransformMacros = macros;
}

export function substituteKnown(text: string, spec: JsonObject) {
  const macros = object(spec.textTransformMacros) ?? {};
  return Object.entries(macros).reduce((result, [name, value]) =>
    typeof value === "string" ? result.replaceAll(`{{${name}}}`, value) : result, text);
}

export function sectionKind(key: string): SillyTavernPresetKind | null {
  return SECTION_KINDS[key] ?? null;
}

function stripView(rule: SillyTavernTextTransformView): TextTransformRuleView {
  const { sourceScope: _scope, inactivePlacements: _inactive, ...runtime } = rule;
  return runtime;
}

function runtimeRule(value: JsonValue): TextTransformRuleView | null {
  const rule = object(value);
  if (!rule || typeof rule.id !== "string" || typeof rule.findRegex !== "string") return null;
  return rule as unknown as TextTransformRuleView;
}

function viewFromRuntime(rule: TextTransformRuleView): SillyTavernTextTransformView {
  const sourceScope = rule.priority < PRIORITY.character ? "global" : rule.priority < PRIORITY.preset ? "character" : "preset";
  return { ...rule, sourceScope, inactivePlacements: [] };
}

File: crates/core/src/compatibility/sillytavern/preset/support.rs
use serde_json::Value;

use crate::{
    graph::{GenerationOptionsIr, ProviderExtensionsIr},
    llm::{
        context::{ContextAssemblyMode, ContextAssemblySpec},
        text_transform::TextTransformRule,
    },
};

use super::super::{
    SillyTavernCompatibilityError, SillyTavernImportWarning, SillyTavernPresetKind,
};

pub(super) struct ImportParts {
    pub context_spec: Option<ContextAssemblySpec>,
    pub generation: Option<GenerationOptionsIr>,
    pub provider_extensions: Option<ProviderExtensionsIr>,
    pub text_transforms: Vec<TextTransformRule>,
    pub inactive_fields: Vec<String>,
    pub warnings: Vec<SillyTavernImportWarning>,
}

impl ImportParts {
    pub fn new(context_spec: Option<ContextAssemblySpec>) -> Self {
        let text_transforms = context_spec
            .as_ref()
            .map(|spec| spec.text_transforms.clone())
            .unwrap_or_default();
        Self {
            context_spec,
            generation: None,
            provider_extensions: None,
            text_transforms,
            inactive_fields: Vec::new(),
            warnings: Vec::new(),
        }
    }
}

pub(super) fn merge_text_transforms(
    target: &mut Vec<TextTransformRule>,
    imported: Vec<TextTransformRule>,
) {
    for rule in imported {
        if let Some(existing) = target.iter_mut().find(|value| value.id == rule.id) {
            *existing = rule;
        } else {
            target.push(rule);
        }
    }
}

pub(super) fn compatibility_error(
    code: &'static str,
    message: impl Into<String>,
) -> SillyTavernCompatibilityError {
    SillyTavernCompatibilityError::new(code, message)
}

pub(super) fn warning(
    warnings: &mut Vec<SillyTavernImportWarning>,
    code: impl Into<String>,
    message: impl Into<String>,
    field: impl Into<Option<String>>,
) {
    warnings.push(SillyTavernImportWarning {
        code: code.into(),
        message: message.into(),
        field: field.into(),
    });
}

pub(super) fn import_name(document: &Value, source_name: Option<&str>) -> String {
    document
        .get("name")
        .and_then(Value::as_str)
        .or(source_name)
        .unwrap_or("SillyTavern import")
        .trim()
        .chars()
        .take(200)
        .collect()
}

pub(super) fn inactive(
    parts: &mut ImportParts,
    field: impl Into<String>,
    message: impl Into<String>,
) {
    let field = field.into();
    parts.inactive_fields.push(field.clone());
    warning(
        &mut parts.warnings,
        "sillytavern_field_inactive",
        message,
        Some(field),
    );
}

pub(super) fn section_kind(key: &str) -> Option<SillyTavernPresetKind> {
    match key {
        "context" => Some(SillyTavernPresetKind::Context),
        "instruct" => Some(SillyTavernPresetKind::Instruct),
        "sysprompt" => Some(SillyTavernPresetKind::SystemPrompt),
        "preset" => Some(SillyTavernPresetKind::TextCompletion),
        "reasoning" => Some(SillyTavernPresetKind::Reasoning),
        _ => None,
    }
}

pub(super) fn empty_context_spec(name: &str) -> ContextAssemblySpec {
    ContextAssemblySpec {
        id: None,
        name: Some(name.into()),
        mode: ContextAssemblyMode::Chat,
        items: Vec::new(),
        budget: None,
        post_process: Vec::new(),
        text_transforms: Vec::new(),
        preview: None,
    }
}
